### PR TITLE
[router]: ensure navigation keeps within closest group

### LINKED ISSUE #29273: Navigating to a shared route doesn't stay within the current group
### Minimal reproducible example

https://github.com/abmenzel/shared-routes

### Which package manager are you using? (Yarn is recommended)

npm

### If the issue is web-related, please select the bundler (`web.bundler` in the `app.json`)

metro

### Summary

I'm trying to set up tabs navigation, where each tab has a nested stack navigator. These will include shared routes that should be accessible from each tab, settings, user etc.
I've following the documentation on [Shared Routes](https://docs.expo.dev/router/advanced/shared-routes/).

#### Setup:
I've used `npx create-expo-app@latest`, then setup the following file-routing structure:

```
(tabs)
-- (home)
---- _layout.tsx
---- index.tsx
---- shared.tsx
-- (profile)
---- _layout.tsx
---- index.tsx
---- shared.tsx
-- _layout.tsx
_layout.tsx
```

From `(home)/index.tsx` and `(profile)/index.tsx` i've set up a `<Link href="/shared">` to navigate to the shared route.

#### Expectation:
When inside a stack navigator (home) or (profile) nested inside (tabs) I expect the links to the shared route to remain in the same stack navigator (home) or (profile).

#### Behaviour:
When on the `(profile)/index` route, clicking the `<Link>` takes you to `(home)/shared.tsx` instead of `(profile)/shared.tsx`.

### Environment

expo-env-info 1.2.0 environment info:
    System:
      OS: macOS 14.4.1
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 20.9.0 - ~/.nvm/versions/node/v20.9.0/bin/node
      Yarn: 1.22.19 - ~/.nvm/versions/node/v20.9.0/bin/yarn
      npm: 10.1.0 - ~/.nvm/versions/node/v20.9.0/bin/npm
    Managers:
      CocoaPods: 1.15.2 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.5, iOS 17.5, macOS 14.5, tvOS 17.5, visionOS 1.2, watchOS 10.5
    IDEs:
      Android Studio: 2022.3 AI-223.8836.35.2231.11090377
      Xcode: 15.4/15F31d - /usr/bin/xcodebuild
    npmPackages:
      expo: ~51.0.9 => 51.0.9
      expo-router: ~3.5.14 => 3.5.14
      react: 18.2.0 => 18.2.0
      react-dom: 18.2.0 => 18.2.0
      react-native: 0.74.1 => 0.74.1
      react-native-web: ~0.19.10 => 0.19.12
    npmGlobalPackages:
      eas-cli: 5.9.0
    Expo Workflow: managed

### LINKED ISSUE #29708: Relative Paths in "index" not working as expected with shared routes (doesn't stay within Stack)
### Minimal reproducible example

https://github.com/samskinner32/sam-expo-router

### Which package manager are you using? (Yarn is recommended)

npm

### Summary

As far as I can tell, relative paths in expo-router are not working as I'd expect when placed within the `index` file and using shared routes. I have a fairly simple router layout where I have three tabs:
1. The starter expo page in tab 1
2. A "shop" stack in tab 2
3. An "account" stack in tab 3

Most screens will be intended to be shared between the "shop" and "account" tabs (e.g. you could view a Product via both screens), so they exist in a shared `(account,shop)` group. Screenshot of this folder layout below:
image

The issue I am running into is that when I attempt to Link to a relative path from within the `(shop)/index` file (e.g. `<Link href="./category">Relative Category</Link>`, it is sending me to `(account)/category` instead of `(shop)/category`. I believe this is related to alphabetical ordering and it deciding on account vs shop.

I would expect the relative path link to stay within the current stack.

If I instead go a step deeper into my stack and do the relative path link from there, it works as expected. For example, if I am on `(shop)/category` to begin with and attempt to link to `./product`, it keeps me within the `(shop)` tab as I expect. So it seems to be an issue specifically with `index` files.

Please let me know if there's anything else I can provide, or if I am making a mistake somewhere.

Thanks!

**TO REPRODUCE**
Simply run the app (it's a basic expo starter project), navigate to the "Shop" tab and press each of the "Relative Category" and "Absolute Category" links and you will see different behavior between the two.

**RELEVANT CODE** in `/app/(tabs)/(shop)/index.tsx`
`<Link href="./category">Relative Category</Link>`
`<Link href="/(shop)/category">Absolute Category</Link>`

### Environment

  expo-env-info 1.2.0 environment info:
    System:
      OS: macOS 14.3
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 21.7.2 - /opt/homebrew/bin/node
      npm: 10.5.0 - /opt/homebrew/bin/npm
      Watchman: 2024.04.01.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.15.2 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.5, iOS 17.5, macOS 14.5, tvOS 17.5, visionOS 1.2, watchOS 10.5
    IDEs:
      Android Studio: 2023.2 AI-232.10300.40.2321.11567975
      Xcode: 15.4/15F31d - /usr/bin/xcodebuild
    npmPackages:
      expo: ~51.0.11 => 51.0.11 
      expo-router: ~3.5.15 => 3.5.15 
      react: 18.2.0 => 18.2.0 
      react-dom: 18.2.0 => 18.2.0 
      react-native: 0.74.2 => 0.74.2 
      react-native-web: ~0.19.10 => 0.19.12 
    npmGlobalPackages:
      eas-cli: 9.2.0
    Expo Workflow: managed

### PR DESCRIPTION
**Pull request by marklawlor**
# Why

Fix #29273, #29708

# How

One of the rules of Expo Router is that navigation should navigation to the closest route, taking the current groups into account. However, there are two scenarios where this did not occur

```
(tabs)
-- (home)
---- _layout.tsx
---- index.tsx
---- shared.tsx
-- (profile)
---- _layout.tsx
---- index.tsx
---- shared.tsx
-- _layout.tsx
_layout.tsx
```

## Bad cached values

If the user navigates like so:

| Route | Notes |
| -- | -- |
| `/` | The initial route |
| `/shared` | `/(home)/shared` |
| `/(profile)` | `/(profile)/index` | 
| `/shared` | Incorrectly goes to `/(home)/shared` |

This incorrect navigation occurs due to `/shared` being a cache key. The solution is simply to removing caching

## Incorect route sorting

Sorting routes should take into account the current route. When the user is at `/(profile)/index` and is navigating via the HREF `/shared` the `/(profile)/shared` should rank higher than `/(home)/shared`

# Test Plan

I believe this behaviour can be fully covered by the unit tests. The linked issue has a test repo.

We did have a unit test that covered this, but I believe that test was incorrect

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-router/src/__tests__/navigation.test.ios.tsx [hunk 1]
```diff
@@ -548,22 +615,22 @@ it('can pop back from a nested modal to a nested sibling', async () => {
...
-  expect(await screen.findByText('found')).toBeOnTheScreen();
+  // expect(screen).toHavePathnameWithParams('/');
+  // expect(await screen.findByText('found')).toBeOnTheScreen();
 
-  act(() => router.push('/123'));
-  expect(await screen.findByText('404')).toBeOnTheScreen();
-  expect(screen).toHavePathname('/123');
-  expect(screen).toHaveSearchParams({
-    'not-found': ['123'],
-  });
```
Comment (EvanBacon) [inline packages/expo-router/src/__tests__/navigation.test.ios.tsx:566]: Is this test not needed anymore?

**Comment by marklawlor** [comment]
- [ ] Bind the value of `getStateFromPath` to the router store to keep the same function signature

File: packages/expo-router/src/fork/getStateFromPath.ts [hunk 2]
```diff
@@ -230,6 +239,25 @@ function sortConfigs(a: RouteConfig, b: RouteConfig): number {
     bParts.push('index');
   }
 
+  // When we navigate, we need to stay within groups as close as possible
+  // Hence, a route is sorted based upon is similiarity to the current state
+  const similarToPreviousA = previousSegments.filter((value, index) => {
+    return value === a.expandedRouteNames[index] && value.startsWith('(') && value.endsWith(')');
```
Comment (EvanBacon) [inline packages/expo-router/src/fork/getStateFromPath.ts:245]: Would prefer better matching on the group names

**Review by EvanBacon — APPROVED** [review]
concerns about context and binding code state to the function are tracked elsewhere.

**Comment by saulamsal** [comment]
any update on this?

**Pull request merged**